nomad/state/paginator/tokenizer.go

```diff
@@ -0,0 +1,84 @@
+package paginator
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Tokenizer is the interface that must be implemented to provide pagination
+// tokens to the Paginator.
+type Tokenizer interface {
+	// GetToken returns the pagination token for the given element.
+	GetToken(interface{}) string
+}
+
+// IDGetter is the interface that must be implemented by structs that need to
+// have their ID as part of the pagination token.
+type IDGetter interface {
+	GetID() string
+}
+
+// NamespaceGetter is the interface that must be implemented by structs that
+// need to have their Namespace as part of the pagination token.
+type NamespaceGetter interface {
+	GetNamespace() string
+}
+
+// CreateIndexGetter is the interface that must be implemented by structs that
+// need to have their CreateIndex as part of the pagination token.
+type CreateIndexGetter interface {
+	GetCreateIndex() uint64
+}
+
+// StructsTokenizerOptions is the configuration provided to a StructsTokenizer.
+type StructsTokenizerOptions struct {
+	WithCreateIndex bool
+	WithNamespace   bool
+	WithID          bool
+}
+
+// StructsTokenizer is an pagination token generator that can create different
+// formats of pagination tokens based on common fields found in the structs
+// package.
+type StructsTokenizer struct {
+	iter Iterator
+	opts StructsTokenizerOptions
+}
+
+// NewStructsTokenizer returns a new StructsTokenizer.
+func NewStructsTokenizer(it Iterator, opts StructsTokenizerOptions) StructsTokenizer {
+	return StructsTokenizer{
+		iter: it,
+		opts: opts,
+	}
+}
+
+func (it StructsTokenizer) GetToken(raw interface{}) string {
```

Comment:
Not a blocker, just thinking: it seems like we could boil this down to a `GetToken() string` interface that is implemented by the types that otherwise implement `GetID`, `GetNamespace`, and/or `GetCreateIndex`. And then we'd just call that, instead of building the same value dynamically via `StructsTokenizer`.

Reply:
Yup, I couldn't find a nice interface since the token format depends on the query params used in the request. The struct doesn't know which format it should use by itself.